ratik/src/main/java/org/opencadc/ratik/ArtifactValidator.java

```diff
@@ -96,15 +120,383 @@ public ArtifactValidator(ArtifactDAO artifactDAO, URI resourceID, boolean trackS
+                    log.info(String.format("Artifact not in L && in R - Explanation4: Artifact in local db but "
+                                               + "siteLocations does not include remote siteID\n"
+                                               + "action: add siteID to Artifact.siteLocations\n"
+                                               + "remote - %s\nremote StorageSite - %s", remote, remoteStorageSite));
+                    //TODO do action
+                    return;
+                }
+            }
+        }
+
+    }
+
+    /**
+     * Artifact in local AND remote
+     *
+     * @param local     The local Artifact.
+     * @param remote    The remote Artifact.
+     */
+    protected void validateLocalAndRemote(Artifact local, Artifact remote) {
+        // discrepancy: artifact.uri in both && artifact.id mismatch (collision)
+        // explanation1: same ID collision due to race condition that metadata-sync has to handle
+        // evidence: no more evidence needed
+        // action: pick winner, create DeletedArtifactEvent for loser, delete loser if it is in L, insert winner if winner was in R
+        log.debug("checking artifact.id mismatch");
+        if (!local.getID().equals(remote.getID())) {
+            log.info(String.format("ID mismatch:\npick winner\nuri - %s\n local - %s %s\nremote - %s %s",
+                                   local.getURI(),
+                                   local.getID(), dateFormat.format(local.getLastModified()),
+                                   remote.getID(), dateFormat.format(remote.getLastModified())));
+            if (local.getContentLastModified().before(remote.getContentLastModified())) {
+                //TODO do action
+            }
+            return;
+        }
+
+        // discrepancy: artifact in both && valid metaChecksum mismatch
+        log.debug("checking valid metaChecksum mismatch");
+        if (!local.getMetaChecksum().equals(remote.getMetaChecksum())) {
+            log.info(String.format("MetaChecksum mismatch:\n uri - %s %s\n local - %s %s\nremote - %s %s",
+                                   local.getURI(), local.getID(), local.getMetaChecksum(),
+                                   dateFormat.format(local.getLastModified()), remote.getMetaChecksum(),
+                                   dateFormat.format(remote.getLastModified())));
+            // explanation1: pending/missed artifact update in L
+            // evidence: ??
+            // action: put Artifact
+            //TODO determine action
+
+            // explanation2: pending/missed artifact update in R
+            // evidence: ??
+            // action: do nothing
+            //TODO determine action
+            return;
+        }
+
+        // Local & remote Artifact's have the same ID.
+        if (local.getID().equals(remote.getID())) {
+            log.info(String.format("Valid: - %s", local));
+        }
+    }
+
+
+    /**
+     * Get a remote Artifact
+     */
```

Comment:
Copied the ArtifactMapper to ArtifactValidator and forgot to remove the original.
I moved the RowMapper classes into org.opencadc.inventory.query in cadc-inventory-util. You need a matching query for the RowMapper so added a method to get the query.
ratik & fenwick use the query and mapper for a TAP query, so added a method that takes a where clause and resourceID, and returns the query results. Could change resourceID to TapClient, since the resourceID doesn't change for a ratik or fenwick instance.
ratik doesn't need the StoreageSite query if it can do a local query, but added it if useful in general? If not will delete. And for completeness could add a class for ObsoleteStorageLocation if that would be useful.

Reply:
This is not what I suggested would be useful and re-usable code. You really only need the ArtifactRowMapper and a string constant with the base query. Everything else is not useful because different applications will do something different. For example, ratik needs to compare artifact streams so it does `order by uri ` while fenwick only does incremental harvesting so it ... does something else.  

cadc-inventory-utll should have ArtifactRowMapper as a normal class. It could have the base query, eg:
```
public static final String BASE_QUERY = "SELECT ... FROM inventory.Artifact";
```
but that does hard-code the schema. I expect that's OK because even though the schema is configurable anywhere (except luskan) I expect that even if it was the solution would be to always use `inventory` in TAP and have luskan re-write the query... because of that it's probably going to be OK to hard code the schema, but it bears considering the consequences.

Same applies to DeletedArtifactEventMapper and DeletedStorageLocationMapper; code is not at the point where I can tell if both fenwick and ratik need a StorageSiteMapper  so leave that in ratik.

Reply:
As above, created RowMapper classes for Artifact, DeletedArtifactEvent, and DeletedStorageLocationEvent, with the base query for each, and updated ratik to use the new classes.